examples/permissions/src/web/app.rs

```diff
@@ -0,0 +1,72 @@
+use std::net::SocketAddr;
+
+use axum::{error_handling::HandleErrorLayer, BoxError};
+use axum_login::{permission_required, AuthManagerLayer};
+use http::StatusCode;
+use sqlx::SqlitePool;
+use time::Duration;
+use tower::ServiceBuilder;
+use tower_sessions::{Expiry, MemoryStore, SessionManagerLayer};
+
+use crate::{
+    users::Backend,
+    web::{auth, protected, restricted},
+};
+
+pub struct App {
+    db: SqlitePool,
+}
+
+impl App {
+    pub async fn new() -> Result<Self, Box<dyn std::error::Error>> {
+        let db = SqlitePool::connect(":memory:").await?;
+        sqlx::migrate!().run(&db).await?;
+
+        Ok(Self { db })
+    }
+
+    pub async fn serve(self) -> Result<(), Box<dyn std::error::Error>> {
+        let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+
+        // Session layer.
+        //
+        // This uses `tower-sessions` to establish a layer that will provide the session
+        // as a request extension.
+        let session_store = MemoryStore::default();
+        let session_layer = SessionManagerLayer::new(session_store)
+            .with_secure(false)
+            .with_expiry(Expiry::OnInactivity(Duration::days(1)));
+
+        // Auth service.
+        //
+        // This combines the session layer with our backend to establish the auth
+        // service which will provide the auth session as a request extension.
+        let backend = Backend::new(self.db);
+        let auth_service = ServiceBuilder::new()
+            .layer(HandleErrorLayer::new(|_: BoxError| async {
+                StatusCode::BAD_REQUEST
+            }))
+            .layer(AuthManagerLayer::new(backend, session_layer));
+
+        let app = restricted::router()
+            .route_layer(permission_required!(
+                Backend,
+                login_url = "/login",
+                "restricted.read",
```

Comment:
I'm not in love with this. While it works, it would be nice to be able to define the groups, say as an enum, and reference them here.

Maybe we need to reconsider the macro, consider another macro, or something else entirely.

Reply:
The permissions will probably be stored as strings on the database level, so either way they will probably be strings.
On the Rust side we can require something that can be converted from/to string (i.e. implements `ToString + 'FromStr' ) and then compare with that, then it can be anything, including an enum.